File: src/Problem2/SolutionTest.java
package Problem2;

import org.junit.Test;

import static org.junit.jupiter.api.Assertions.*;

public class SolutionTest {
    @Test
    public void test1() {
        Solution solution = new Solution();
        ListNode n1 = convertArrayToList(new int[] { 2,4,3 });
        ListNode n2 = convertArrayToList(new int[] { 5,6,4 });
        ListNode actual = solution.addTwoNumbers(n1, n2);
        ListNode expected = convertArrayToList(new int[] { 7,0,8 });
        assertListsMatch(expected, actual);
    }

    @Test
    public void test2() {
        Solution solution = new Solution();
        ListNode n1 = convertArrayToList(new int[] { 7,2,4,3 });
        ListNode n2 = convertArrayToList(new int[] { 5,6,4 });
        ListNode actual = solution.addTwoNumbers(n1, n2);
        ListNode expected = convertArrayToList(new int[] { 2,9,8,3 });
        assertListsMatch(expected, actual);
    }

    @Test
    public void test3() {
        Solution solution = new Solution();
        ListNode n1 = convertArrayToList(new int[] { 0 });
        ListNode n2 = convertArrayToList(new int[] { 0 });
        ListNode actual = solution.addTwoNumbers(n1, n2);
        ListNode expected = convertArrayToList(new int[] { 0 });
        assertListsMatch(expected, actual);
    }

    @Test
    public void test4() {
        Solution solution = new Solution();
        ListNode n1 = convertArrayToList(new int[] { 9,9,9,9,9,9,9 });
        ListNode n2 = convertArrayToList(new int[] { 9,9,9,9 });
        ListNode actual = solution.addTwoNumbers(n1, n2);
        ListNode expected = convertArrayToList(new int[] { 8,9,9,9,0,0,0,1 });
        assertListsMatch(expected, actual);
    }

    @Test
    public void test5() {
        Solution solution = new Solution();
        ListNode n1 = convertArrayToList(new int[] { 9 });
        ListNode n2 = convertArrayToList(new int[] { 9 });
        ListNode actual = solution.addTwoNumbers(n1, n2);
        ListNode expected = convertArrayToList(new int[] { 8,1 });
        assertListsMatch(expected, actual);
    }

    private ListNode convertArrayToList(int[] arrayOfInts) {
        ListNode previousNode = null;
        ListNode result = null;

        for (int i = arrayOfInts.length - 1; i >= 0; i--) {
            result = new ListNode(arrayOfInts[i], previousNode);
            previousNode = result;
        }

        return result;
    }

    private void  assertListsMatch(ListNode expected, ListNode actual) {
        ListNode nextExpected = expected;
        ListNode nextActual = actual;

        while ((nextExpected != null) && (nextActual != null)) {
            assertEquals(nextExpected.val, nextActual.val);
            nextExpected = nextExpected.next;
            nextActual = nextActual.next;
        }

        assertFalse((nextExpected != null) || (nextActual != null));
    }
}

File: src/problem21/SolutionTest.kt
package problem21

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class SolutionTest {
    @Test
    public fun test1() {
        val list1 = convertArrayToList(arrayOf());
        val list2 = convertArrayToList(arrayOf())

        val solution = Solution()
        val mergedList = solution.mergeTwoLists(list1, list2)

        val expectedResult = convertArrayToList(arrayOf())
        assertListsMatch(expectedResult, mergedList)
    }

    @Test
    public fun test2() {
        val list1 = convertArrayToList(arrayOf(1, 2, 4));
        val list2 = convertArrayToList(arrayOf())

        val solution = Solution()
        val mergedList = solution.mergeTwoLists(list1, list2)

        val expectedResult = convertArrayToList(arrayOf(1,2,4))
        assertListsMatch(expectedResult, mergedList)
    }

    @Test
    public fun test3() {
        val list1 = convertArrayToList(arrayOf());
        val list2 = convertArrayToList(arrayOf(1, 3, 4));

        val solution = Solution()
        val mergedList = solution.mergeTwoLists(list1, list2)

        val expectedResult = convertArrayToList(arrayOf(1,3,4))
        assertListsMatch(expectedResult, mergedList)
    }

    @Test
    public fun test4() {
        val list1 = convertArrayToList(arrayOf(1, 2, 4));
        val list2 = convertArrayToList(arrayOf(1, 3, 4));

        val solution = Solution()
        val mergedList = solution.mergeTwoLists(list1, list2)

        val expectedResult = convertArrayToList(arrayOf(1,1,2,3,4,4))
        assertListsMatch(expectedResult, mergedList)
    }

    @Test
    public fun test5() {
        val list1 = convertArrayToList(arrayOf(1, 2, 100, 150, 9090, 54930));
        val list2 = convertArrayToList(arrayOf(1, 3, 4, 100, 110, 359, 100000));

        val solution = Solution()
        val mergedList = solution.mergeTwoLists(list1, list2)

        val expectedResult = convertArrayToList(arrayOf(1,1,2,3,4,100,100,110,150,359,9090,54930,100000))
        assertListsMatch(expectedResult, mergedList)
    }

    fun convertArrayToList(arrayOfInts: Array<Int>): ListNode? {
        var previousNode: ListNode? = null
        var result: ListNode? = null

        arrayOfInts.reversedArray().forEach { i ->
            result = ListNode(i, previousNode)
            previousNode = result
        }

        return result
    }

    fun assertListsMatch(expected: ListNode?, actual: ListNode?) {
        var nextExpected: ListNode? = expected
        var nextActual: ListNode? = actual

        while ((nextExpected != null) && (nextActual != null)) {
            assertEquals(nextExpected.`val`, nextActual.`val`)
            nextExpected = nextExpected.next
            nextActual = nextActual.next
        }

        assertFalse((nextExpected != null) || (nextActual != null))
    }
}